File: decorator-example/python/Decorator.py
class Donut:
    def cost(self):
        return 5

class ChocolateDecorator:
    def __init__(self, coffee):
        self.coffee = coffee

    def cost(self):
        return self.coffee.cost() + 2

class CaramelDecorator:
    def __init__(self, coffee):
        self.coffee = coffee

    def cost(self):
        return self.coffee.cost() + 3

simple_coffee = Donut()
print('Simple donut: $' + str(simple_coffee.cost()))

coffee_with_milk = ChocolateDecorator(simple_coffee)
print('Chocolate donut: $' + str(coffee_with_milk.cost()))

coffee_with_milk_and_caramel = CaramelDecorator(coffee_with_milk)
print('Chocolate caramel donut: $' + str(coffee_with_milk_and_caramel.cost()))

File: decorator-example/js/Decorator.js
class Donut {
  cost() {
    return 5;
  }
}

class ChocolateDecorator {
  constructor(coffee) {
    this.coffee = coffee;
  }

  cost() {
    return this.coffee.cost() + 2;
  }
}

class CaramelDecorator {
  constructor(coffee) {
    this.coffee = coffee;
  }

  cost() {
    return this.coffee.cost() + 3;
  }
}

const simpleCoffee = new Donut();
console.log('Simple donut: $' + simpleCoffee.cost());

const coffeeWithMilk = new ChocolateDecorator(simpleCoffee);
console.log('Chocolate donut: $' + coffeeWithMilk.cost());

const coffeeWithMilkAndCaramel = new CaramelDecorator(coffeeWithMilk);
console.log('Chocolate caramel donut: $' + coffeeWithMilkAndCaramel.cost());
